pkg/kubelet/userns_manager.go

```diff
@@ -0,0 +1,600 @@
+	return err
+}
+
+// isSet checks if the specified index is already set.
+func (m *usernsManager) isSet(v uint32) bool {
+	index := v / userNsLength
+	return m.used.isSet(index)
+}
+
+// allocateOne finds a free user namespace and allocate it to the specified pod.
+// The first return value is the first ID in the user namespace, the second returns
+// the length for the user namespace range.
+func (m *usernsManager) allocateOne(pod string) (firstID uint32, length uint32, err error) {
+	if m.numAllocated >= maxPods {
+		return 0, 0, fmt.Errorf("limit on count of pods with user namespaces exceeded (limit is %v, current pods with userns: %v)", maxPods, m.numAllocated)
+	}
+	m.numAllocated++
+	defer func() {
+		if err != nil {
+			m.numAllocated--
+		}
+	}()
+
+	firstZero, found := m.used.findAvailable()
+	if !found {
+		return 0, 0, fmt.Errorf("could not find an empty slot to allocate a user namespace")
+	}
+
+	klog.V(5).InfoS("new pod user namespace allocation", "podUID", pod)
+
+	firstID = firstZero * userNsLength
+	m.usedBy[pod] = firstID
+	return firstID, userNsLength, nil
+}
+
+// record stores the user namespace [from; from+length] to the specified pod.
+func (m *usernsManager) record(pod types.UID, from, length uint32) (err error) {
+	if length != userNsLength {
+		return fmt.Errorf("wrong user namespace length %v", length)
+	}
+	if from%userNsLength != 0 {
+		return fmt.Errorf("wrong user namespace offset specified %v", from)
+	}
+	prevFrom, found := m.usedBy[string(pod)]
+	if found && prevFrom != from {
+		return fmt.Errorf("different user namespace range already used by pod %q", pod)
+	}
+	index := from / userNsLength
+	// if the pod wasn't found then verify the range is free.
+	if !found && m.used.isSet(index) {
+		return fmt.Errorf("range picked for pod %q already taken", pod)
+	}
+	// The pod is already registered, nothing to do.
+	if found && prevFrom == from {
+		return nil
+	}
+	if m.numAllocated >= maxPods {
+		return fmt.Errorf("limit on count of pods with user namespaces exceeded (limit is %v, current pods with userns: %v)", maxPods, m.numAllocated)
+	}
+	m.numAllocated++
+	defer func() {
+		if err != nil {
+			m.numAllocated--
+		}
+	}()
+
+	klog.V(5).InfoS("new pod user namespace allocation", "podUID", pod)
+
+	// "from" is a ID (UID/GID), set the corresponding userns of size
+	// userNsLength in the bit-array.
+	m.used.set(index)
+	m.usedBy[string(pod)] = from
+	return nil
+}
+
+// Release releases the user namespace allocated to the specified pod.
+func (m *usernsManager) Release(podUID types.UID) {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.UserNamespacesSupport) {
+		return
+	}
+
+	m.Lock()
+	defer m.Unlock()
+
+	m.releaseWithLock(podUID)
+}
+
+func (m *usernsManager) releaseWithLock(podUID types.UID) {
+	pod := string(podUID)
+	v, ok := m.usedBy[pod]
+	if !ok {
+		klog.V(5).InfoS("pod user namespace allocation not present", "podUID", podUID)
+		return
+	}
+	delete(m.usedBy, pod)
```

Comment:
i would have expected to see a delete operation on the user mappings file in this operation as well.

am i missing some other way content on disk was being cleared?

Reply:
Yes and no, we were relying that the pod folder is deleted. But Giuseppe already updated this to delete the file :)